examples/hotrod/pkg/pool/pool_test.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pool
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+var n int32 = 0
+var wg sync.WaitGroup
+
+func testJob() {
+	atomic.AddInt32(&n, 1)
+	time.Sleep(time.Second)
+	wg.Done()
+}
+
+func TestExecute(t *testing.T) {
+	tests := []struct {
+		name   string
+		pool   *Pool
+		job    func()
+		used   time.Duration
+		expect int32
+	}{
+		{
+			name:   "1 worker, loop 1",
+			pool:   New(1),
+			job:    testJob,
+			used:   time.Second + 500000000,
+			expect: 1,
+		},
+		{
+			name:   "2 worker, loop 2",
+			pool:   New(2),
+			job:    testJob,
+			used:   time.Second + 500000000,
+			expect: 2,
+		},
+		{
+			name:   "3 worker, loop 3",
+			pool:   New(3),
+			job:    testJob,
+			used:   time.Second + 500000000,
+			expect: 3,
+		},
+		{
+			name:   "4 worker, loop 4",
+			pool:   New(4),
+			job:    testJob,
+			used:   time.Second + 500000000,
+			expect: 4,
+		},
+		{
+			name:   "1 worker, loop 5",
+			pool:   New(1),
+			job:    testJob,
+			used:   time.Second*5 + 500000000,
+			expect: 5,
+		},
+		{
+			name:   "2 worker, loop 6",
+			pool:   New(2),
+			job:    testJob,
+			used:   time.Second*3 + 500000000,
+			expect: 6,
+		},
+	}
+	for index, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			wg.Add(index + 1)
```

Comment:
This looks like incorrect usage of wg.Add - since you're only adding one job, it should be Add(1).

Reply:
the job will be executed (index + 1) times

Reply:
This is difficult to reason about, and makes the whole test fragile by creating dependencies on the order of tests. If you need N job runs per test, then you can make N an explicit parameter of the test (a field in the struct).

Also, if you're running jobs in a loop, it's still easier to reason about them if you call wg.Add(1) before each job, rather than wg.Add(N).